ethereum/sidecar/server.go

```diff
@@ -428,6 +537,342 @@ func (s *Server) startGRPCServer(
+				seqStart.String(),
+				seqEnd.String(),
+				err,
+			)
+		}
+
+		// We know how many events we should receive. Any difference suggests
+		// an error on `mezod` server side (e.g. gaps in stored event database).
+		expectedLength := seqEnd.Sub(seqStart).Int64()
+		if int64(len(events)) != expectedLength {
+			return nil, fmt.Errorf(
+				"fetched unexpected number of AssetsUnlocked events for "+
+					"range [%s, %s); expected %d, got %d",
+				seqStart.String(),
+				seqEnd.String(),
+				expectedLength,
+				len(events),
+			)
+		}
+
+		for i := len(events) - 1; i >= 0; i-- {
+			event := events[i]
+			if event.BlockTime < cutOffBlockTime {
+				// We found an event older than cut-off time.
+				break outer
+			}
+			recentEvents = append(recentEvents, event)
+		}
+
+		// If the length of fetched events was smaller than the batch size,
+		// there are no more events to fetch.
+		if len(events) < s.assetsUnlockedBatchSize {
+			break
+		}
+
+		sequenceTip = seqStart.SubRaw(1)
+	}
+
+	// Ensure events are sorted in ascending order of their unlock sequences.
+	sort.Slice(recentEvents, func(i, j int) bool {
+		return recentEvents[i].UnlockSequence.LT(recentEvents[j].UnlockSequence)
+	})
+
+	return recentEvents, nil
+}
+
+// findUnattestedAssetsUnlockedEvents finds unattested `AssetsUnlocked` events
+// from the passed input events. An event is considered attested if there
+// was an `AssetsUnlockConfirmed` event emitted for it on Ethereum and its
+// unlock sequence is among the confirmed unlocks.
+func (s *Server) findUnattestedAssetsUnlockedEvents(
+	ctx context.Context,
+	inputEvents []bridgetypes.AssetsUnlockedEvent,
+) ([]bridgetypes.AssetsUnlockedEvent, error) {
+	// If there are no input events, return early with an empty list.
+	if len(inputEvents) == 0 {
+		return []bridgetypes.AssetsUnlockedEvent{}, nil
+	}
+
+	// Use unlock sequences of the input events as a filter.
+	unlockSequences := make([]*big.Int, len(inputEvents))
+	for i, event := range inputEvents {
+		unlockSequences[i] = event.UnlockSequence.BigInt()
+	}
+
+	// Finalized block is considered safe from reorgs as it is 64-96 blocks
+	// behind the current tip. It can be used as a filter.
+	finalizedBlock, err := s.chain.FinalizedBlock(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the finalized block: [%w]", err)
+	}
+	endBlock := finalizedBlock.Uint64()
+
+	// Search for `AssetsUnlockConfirmed` events for the given unlock sequences.
+	// Limit the search to finalized blocks.
+	confirmedEvents, err := s.bridgeContract.PastAssetsUnlockConfirmedEvents(
+		0, // TODO: Would setting startBlock to a lower value improve performance?
+		&endBlock,
+		unlockSequences,
+		nil,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to fetch past AssetsUnlockConfirmed events: [%w]",
+			err,
+		)
+	}
+
+	// TODO: We might need to split getting PastAssetsUnlockConfirmedEvents into batches.
+
+	attestedUnlockSequences := make(map[string]bool)
```

Comment:
Let's use the word "confirmed". Attestation is a short-living process during which validators provide their attestations. Even gathering enough attestations is marked as confirmed.

Reply:
Done in 0d2144781d4d683df2c171302b0c640049659133.